Harmonize watchers endpoints

NobodysNightmare opened the pull request:
## OpenProject Work Packages
- https://community.openproject.org/work_packages/17755
- includes non WP specifications that harmonize the API
## Content
- new endpoint for current watchers of a work package
- clarification on when to return 404 and 403 for some endpoints
  - this change _should_ not affect the actual implementation, however I did not verify that
- allow adding single watchers
- ~~endpoint for available watchers is now also placed under a project scope~~
## Changes to existing endpoints
- [x] new error identifier that shall also be used for `PATCH work package` (and its form)
- [x] there is an endpoint to list watchers of a work package
- [x] watchers are linked in Work Package
- [x] the `watch` link of the work package has an updated `data` section (that is now called `payload`)
- [x] watchers are embedded as collection resource
- [x] add watchers now accepts a new request format
  - [x] adapt angular client
- [x] ~~add watchers now returns the list of current watchers~~
## Additional changes required to implementation

The implementation thankfully differed from the existing documentaion already. Here are the things I could find:
- [x] `watchChanges` has to be called `watch` (same for `unwatchChanges` which is now `unwatch`)
- [x] fix HTTP 500 adding a non-existant user as watcher
- [x] check response when removing a non-watching user/non existant user as watcher
## Additional changes required for documentation
- [x] move available watchers endpoint back to WP scope

:warning: This PR should be merged into `release/4.2`

NobodysNightmare committed:
specify endpoint for current watchers

[ci skip]

TeatroIO commented:
I've prepared a stage. <a href='http://fix-specify-watchers-link.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Click to open.</a>

NobodysNightmare commented on doc/apiv3-documentation.apib:
```diff
@@ -2207,6 +2207,7 @@ Note that due to sharing this might be more than the versions *defined* by that
 | timeEntries        | All time entries logged on the work package. Please note that this is a link to an HTML resource for now and as such, the link is subject to change.  | N/A          |              | READ                  | **Permission** view time entries          |
 | type               | The type of the work package                                                                                                                          | Type         | not null     | READ / WRITE          |                                           |
 | version            | The version associated to the work package                                                                                                            | Version      |              | READ / WRITE          |                                           |
+| watchers           | All users that are currently watching this work package                                                                                               | Users        |              | READ                  |                                           |
```
The question arises to me if everyone can see the list of watchers or if special privileges are neccessary. Which?

NobodysNightmare committed:
require permissions to see watchers

[ci skip]

NobodysNightmare committed:
clarify 403/404 for WP scoped resources

[ci skip]

NobodysNightmare committed:
change the way watchers are added

- using links
- possibility for multiple watchers at a time
- always returns all watchers (pagination stilll possible)

Note that the term "link object" is introduced by an as of now pending
PR

[ci skip]

NobodysNightmare committed:
specify further error cases

[ci skip]

NobodysNightmare committed:
fix example link

[ci skip]

NobodysNightmare committed:
bring available watchers in line with assignees etc

- endpoint also scoped by project
- thus also returning all possible watchers, not only the ones "not
watching"

[ci skip]

NobodysNightmare committed:
clarification on 403/404 of available xxx

[ci skip]

ulferts commented on doc/apiv3-documentation.apib:
```diff
@@ -2296,14 +2297,14 @@ Note that due to sharing this might be more than the versions *defined* by that
                     "watch": {
                         "href": "/api/v3/work_packages/1528/watchers",
                         "method": "post",
                         "data": {
```
As I understand this, all those attributes in the link section apart from "href" are not covered by any standard. As such I propose not using "data" but "payload", so that we have consistency with our forms.

ulferts replied:
BTW, should we, meaning you @NobodysNightmare, add a section about which attributes to expect in the links section?

NobodysNightmare replied:
https://github.com/opf/openproject/pull/2327 already adds such a section.
We, meaning I, should add the `payload` there.

Indeed I was thinking about a good name, but couldn't come up with something as beautiful as `payload`.
I will use that.

ulferts commented on doc/apiv3-documentation.apib:
```diff
@@ -2444,6 +2462,20 @@ The value of `lockVersion` is used to implement [optimistic locking](http://en.w
+
+    Returned if the work package does not exist or the client does not have sufficient permissions to see it.
+
+    **Required permission:** view work package
```
this should be `edit work package`

NobodysNightmare replied:
No it shouldn't. This is the trick with all this 403/404 clarification we can return proper status codes when the user is allowed to know about them.

OMG, my ASCII-Art skills:

```
           | can edit | can't edit |
-----------|----------|------------|
can view   |  200     |   403      |
can't view |  404     |   404      |

doesn't exist: 404
```

In other words: If you are _not allowed to see_ the WP, you will **always** receive a 404. If you are _allowed to see_ the WP, you will receive a 403 if you are _solely_ missing the **edit** rights.

Obviously 404 is also possible for WPs that are _really_ non existent.

ulferts commented on doc/apiv3-documentation.apib:
```diff
@@ -2490,43 +2522,219 @@ For more details and all possible responses see the general specification of [Fo
+
+    **Required permission:** edit work package
+
+    *Note that you will only receive this error, if you are at least allowed to see the corresponding work package.*
```
This phrase is hard to understand. My understanding is that you need the `view work package` permission in order to get the 403 on a work package for which you are lacking the `edit work package` permission. Otherwise a 404 is returned. Correct?

NobodysNightmare replied:
yep

ulferts commented on doc/apiv3-documentation.apib:
```diff
@@ -2490,43 +2522,219 @@ For more details and all possible responses see the general specification of [Fo
+
+    Returned if the work package does not exist or the client does not have sufficient permissions to see it.
+
+    **Required permission:** view work package
```
please see above

ulferts commented on doc/apiv3-documentation.apib:
```diff
@@ -2490,43 +2522,219 @@ For more details and all possible responses see the general specification of [Fo
+            {
+                "_type": "Error",
+                "errorIdentifier": "urn:openproject-org:api:v3:errors:InvalidRequestBody",
+                "message": "The request body was not a single JSON object."
```
This message does not cover the whole of the examples mentioned above.

NobodysNightmare replied:
First of all: It is an example. All rules for error messages are defined in the errors section, the rest is up to the implementation.

However, implementors are free to return an apropriate message for each scenario (I'd like that) **or** return a single message apropriate for _every_ scenario (as currently done in most cases).

ulferts commented on doc/apiv3-documentation.apib:
```diff
@@ -2490,43 +2522,219 @@ For more details and all possible responses see the general specification of [Fo
+                "_type": "Error",
+                "errorIdentifier": "urn:openproject-org:api:v3:errors:PropertyConstraintViolation",
+                "message": "Expected resources of type 'User', but got a 'Status'.",
+            }
```
What about the following cases (in both, the user making the request has the necessary permissions):
- One user is specified to become watcher and he is not allowed to be a watcher (not included in list of available watchers)
- Multiple users are specified to become watchers and some of them are not allowed to become watchers (not included in the list of available watchers)

ulferts commented on doc/apiv3-documentation.apib:
```diff
@@ -2538,14 +2746,46 @@ For more details and all possible responses see the general specification of [Fo
+                "errorIdentifier": "urn:openproject-org:api:v3:errors:NotFound",
+                "message": "The specified work package does not exist."
+            }
+
```
I now begin to wonder if we should really start allowing setting multiple watchers at once. Where does that stop? We now need to also be able to remove multiple watchers. Maybe not from the angular client but it would be consistent. But the ability for bulk operations might also be useful on other resources. Can we get away with only allowing this for assignments, and excluding creation and manipulation?

NobodysNightmare replied:
I tend to agree...

Lets postpone a decision until monday to be able to think about it again.

NobodysNightmare replied:
done that

NobodysNightmare committed:
also return 422 for non-watchable users

[ci skip]

NobodysNightmare commented:
@ulferts I added users that are not allowed as watchers to the list of 422 reasons and clarified that either all or no users are added as watchers.

**Open Question**: Should we add another error identifier for any of those cases? Providing a wrong resource type (status when user was expected) sounds like something we will have to deal with very often and thus might answer quite specifically... **edit: I said yes**

NobodysNightmare committed:
Merge remote-tracking branch 'origin/dev' into fix/specify_watchers_link

Conflicts:
	doc/apiv3-documentation.apib

[ci skip]

NobodysNightmare committed:
rename data to payload, reference it in "links"

[ci skip]

NobodysNightmare committed:
only allow to add one watcher at a time

NobodysNightmare committed:
specify error id for type mismatch of links

[ci skip]

NobodysNightmare committed:
fixed example

[ci skip]

NobodysNightmare commented:
One point found during a third party review:
- [x] Specification should state the response for trying to add/remove a non existant user

myabc commented:
Rebase me!

NobodysNightmare committed:
Merge branch 'release/4.2' into fix/specify_watchers_link

Conflicts:
	doc/apiv3-documentation.apib

NobodysNightmare committed:
return ResourceTypeMismatch for wrong resources

- this is according to new specification changes
- I have no clue why the initial author did this "wrap into a validation error" business

NobodysNightmare committed:
adapt watch/unwatch links to specification

- they are called (un)watch, not (un)watchChanges
- watch link indicates the required payload using the payload attribute
- required payload already indicated using new (as of this commit invalid) format

NobodysNightmare committed:
remove redundant authentication

already performed on project level

NobodysNightmare committed:
style resource link parser

NobodysNightmare committed:
update roar version

NobodysNightmare committed:
extract parsable link into own representer

NobodysNightmare committed:
accept new request format for add watcher

also adapt the angular frontend to speak the new format

NobodysNightmare committed:
avoid useless JSON roundtrip

know your library they said...

NobodysNightmare committed:
fix specs to respect new error identifier

NobodysNightmare commented on doc/apiv3-documentation.apib:
```diff
@@ -3068,22 +3107,238 @@ For more details and all possible responses see the general specification of [Fo
+
++ Response 200 (application/hal+json)
+
+    [Watchers][]
```
We defined to return the collection of current watchers in response to a successful add watcher call.

However, I just realized that to add a watcher it suffices to have the permission _view work package_ suffices (for self). But to list watchers of a work package, it is neccessary to have to _view work package watchers_ permission.

This is a conflict, because by adding self to the watchers I can effectively list all other watchers.

@ulferts Any thoughts? I'd currently say we simply return the user that was just added (current behaviour)

ulferts replied:
@NobodysNightmare I am fine with your proposal. It also seems to be more consistent with the rest of the API.

NobodysNightmare committed:
clarify more edge cases regarding watchers

NobodysNightmare commented:
@ulferts We decided (some time ago), that it would make sense from an API perspective to have the available watchers on the same level as the available assignees and available responsibles. That is: They should all reside under `projects/:id`.

However I noted, that on the ruby side I can't ask a project for its available watchers. I can only do that for an `acts_as_watchable`.

Should I:

a) Move the available watchers endpoint back to `work_packages/:id`
b) Allow the project on the ruby side to know its possible watchers (effectively moving the method in the backend)

What is your opinion here? From a code perspective I'd favour a), from an API perspective I'd favour b).

NobodysNightmare committed:
try to improve readability of watcher spec

incl. Rubocop fixes

ulferts commented:
@NobodysNightmare as there are different things one can watch (WP, wiki, ...) and those require separate permissions, the list can not be on the project scope. It therefore has to be moved to the work package scope IMO.

NobodysNightmare committed:
allow to view list of current watchers

NobodysNightmare committed:
fix HTTP 500 in case of not allowed watcher

NobodysNightmare committed:
also use named parameters for RemoveWatcher service

NobodysNightmare committed:
specify and test edge cases for watcher deletion

reconsidered when to return 404:

while the watcher resource does technically not exist when a user is not watching,
it seems more client friendly to still return a success code even if nothing was done

Still I differentiated that from user that does not exist at all

NobodysNightmare committed:
further specification refinements

- remove build warnings
- move available watchers endpoint back to WP scope

NobodysNightmare committed:
correctly link & embed watchers in WP

NobodysNightmare commented:
Considering this done, even though I saw things I did not want to see...
But those have to wait for a later PR.

Waiting for travis, if he is cool this is ready for review.

edit: Travis is looking kind of good...

NobodysNightmare committed:
add spec for new link object

ulferts commented on Gemfile.lock:
```diff
@@ -363,12 +363,12 @@ GEM
       uber (~> 0.0.7)
     request_store (1.1.0)
-    roar (1.0.0)
+    roar (1.0.1)
```
@NobodysNightmare nothing against updating the gems but is the update require for the PR?

NobodysNightmare replied:
Somehow I typed the characters to update this gem and now it is updated. I don't really recall if this was neccessary or not ^^

ulferts commented:
@NobodysNightmare is the targeted branche (`dev`) still correct? We should probably reuse part of it in 4.2, right?

NobodysNightmare commented:
right

NobodysNightmare commented:
I will reopen this PR against the `release/4.2` branch
